enclave/src/main/java/org/hyperledger/besu/enclave/Enclave.java

```
@@ -26,101 +28,117 @@
+      return result.get();
     } catch (final JsonProcessingException e) {
       throw new EnclaveException("Failed to serialise request to json body.");
+    } catch (final ExecutionException e) {
+      if (e.getCause() instanceof EnclaveException) {
+        throw (EnclaveException) e.getCause();
+      }
+      throw new EnclaveException("Failure during reception", e);
+    } catch (final InterruptedException e) {
+      throw new EnclaveException("Task interrupted while waiting for Enclave response.", e);
+    } catch (final Exception e) {
+      throw new EnclaveException("Other error", e);
     }
-    final String url = enclaveUri.resolve(endpoint).toString();
-    return new Request.Builder().url(url).post(body).build();
   }
 
-  private <T> T executePost(final Request request, final Class<T> responseType) {
-
-    final Response response;
-    final String responseBody;
-    try {
-      response = client.newCall(request).execute();
-      responseBody = response.body().string();
-    } catch (final IOException e) {
-      throw new RuntimeException("Failed to contact Enclave", e);
-    }
-
-    try {
-      if (response.isSuccessful()) {
-        return objectMapper.readValue(responseBody, responseType);
-      } else {
-        final ErrorResponse errorResponse =
-            objectMapper.readValue(responseBody, ErrorResponse.class);
-        throw new EnclaveException(errorResponse.getError());
-      }
-    } catch (final JsonParseException | JsonMappingException e) {
-      throw new EnclaveException("Failed to deserialise received json", e);
-    } catch (final IOException e) {
-      throw new EnclaveException("Decoding json stream failed.", e);
-    }
+  private <T> void handleResponse(
+      final HttpClientResponse response,
+      final Class<T> responseType,
+      final CompletableFuture<T> future) {
+    response.bodyHandler(
+        responseBody -> {
+          try {
+            if (response.statusCode() == 200) {
```

Do we want to be more nuanced than this? Orion will return a 404 if the key can't be found, do we want to throw an exception in that case?

---

I think this replicates the behaviour of the previous implementation, which was that a 404 would present as an EnclaveException (containing response's body)

---

Yes, it does replicate the old behaviour.